lib/trento/domain/sap_system/sap_system.ex

```diff
@@ -290,6 +297,188 @@ defmodule Trento.Domain.SapSystem do
     }
   end
 
+  # Deregister an application instance and emit a ApplicationInstanceDeregistered
+  # also emit SapSystemDeregistered event if this was the last application instance
+  def execute(
+        %SapSystem{
+          sap_system_id: sap_system_id,
+          application: %Application{instances: instances}
+        } = sap_system,
+        %DeregisterApplicationInstance{
+          instance_number: instance_number,
+          sap_system_id: sap_system_id,
+          host_id: host_id,
+          deregistered_at: deregistered_at
+        } = command
+      ) do
+    instance = get_instance(instances, host_id, instance_number)
```

Comment:
For me personally this function looks really non declarative and confusing.

In my view this should look like:
```
    sap_system
    |> Multi.new()
    |> Multi.execute(fn _ ->
      %ApplicationInstanceDeregistered{
        sap_system_id: sap_system_id,
        ...
      }
    |> Multi.execute(fn sap_system ->
      maybe_emit_sap_system_deregistered_events(
        sap_system,
        deregistered_at
      )
```

And that's all. 
You check all the logic about what to do with the sap system at `maybe_emit_sap_system_deregistered_events`, where you check if instances of application or database are nil, or there is not any instance with `ascs/pas` type.
It is much easier to check what's in the current state rather than in the incoming command.

It would simplify a lot and make the code meaningful